### PR TITLE
Upgrade Jackson to 2.12

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR upgrades all Jackson dependencies from 2.11 to 2.12. 

In the newest version, Jackson made a few changes to how XML serialization and deserialization handled inferring types and handling default scenarios. Default handling for empty lists and strings changed from null to being empty, additionally Jackson introduced a new concept of [Coercion ](https://cowtowncoder.medium.com/jackson-2-12-most-wanted-4-5-cbc91c00bcd2) which takes precedence over configurations such as `FromXmlParser.Feature.EMPTY_ELEMENT_AS_NULL`.

### CODE REVIEW

**Review by srnagar — APPROVED** [review]
LGTM!

File: sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java [hunk 1]
```diff
@@ -234,96 +263,178 @@ public String serializeList(List<?> list, CollectionFormat format) {
...
+                }
+
+                return false;
+            } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException ignored) {
+                return false;
+            }
+        }
+
+        private String getPotentialSetterName() {
+            String fieldName = declaringField.getName();
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java:435]: Is this guaranteed to be non-null and non-empty?

File: sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java [hunk 2]
```diff
@@ -234,96 +263,178 @@ public String serializeList(List<?> list, CollectionFormat format) {
...
+            this.values = new HashMap<>();
+            this.declaringField = declaringField;
+        }
+
+        boolean headerStartsWithPrefix(String headerName) {
+            return headerName.startsWith(prefix);
+        }
+
+        void addHeader(String headerName, String headerValue) {
+            values.put(headerName.substring(prefixLength), headerValue);
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java:377]: Should there be a check for `headerName.length() >= prefixLength`?

File: sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java [hunk 3]
```diff
@@ -67,30 +78,41 @@
...
+        this.headerMapper = initializeMapperBuilder(JsonMapper.builder())
+            .enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_VALUES)
+            .build();
+
+        this.xmlMapper = initializeMapperBuilder(XmlMapper.builder())
+            .defaultUseWrapper(false)
+            .enable(ToXmlGenerator.Feature.WRITE_XML_DECLARATION)
+            /*
+             * In Jackson 2.12 the default value of this feature changed from true to false.
+             * https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.12#xml-module
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java:99]: Agreed, but at least it is called out and easily worked around.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java:433]: Should be, this is a field retrieved using the `getDeclaredFields()` method on `Class`, so I would be surprised if a null or empty name was returned

File: sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java:375]: This is already checked for this by the caller, validating that the header name `startsWith` the prefix.

**Comment by alzimmermsft** [comment]
@kushagraThapar could you review, or find someone to review, the Cosmos changes?

**Review by conniey — APPROVED** [review]

**Comment by kushagraThapar** [comment]
> @kushagraThapar could you review, or find someone to review, the Cosmos changes?

I don't see any cosmos code changes. And since you already ran the cosmos ci, I think you are good from cosmos point of view.

**Review by kushagraThapar — APPROVED** [review]
@alzimmermsft - if you have time and bandwidth, please run the cosmos live tests as well, just to give us good confidence in this upgrade.

**Review by gapra-msft — COMMENTED** [review]
Looks good from a storage perspective

**Pull request merged**